consensus/src/sync/live/block_queue/block_request_component.rs

```diff
@@ -170,55 +170,72 @@ impl<N: Network> BlockRequestComponent<N> {
                                 has_body = block.body().is_some(),
                                 include_body = request.include_body,
                                 "Received block with body where none was expected or vice versa",
                             );
+                                return false;
+                            }
+                        }
+
+                        // Checks that the first block's parent was part of the block locators.
+                        let first_block = blocks.first().unwrap(); // cannot be empty
+                        if !request.locators.contains(first_block.parent_hash()) {
+                            log::error!("Received invalid chain of missing blocks (first block's parent not in block locators)");
                             return false;
                         }
-                    }
 
-                    // Checks that the first block's parent was part of the block locators.
-                    let first_block = blocks.first().unwrap(); // cannot be empty
-                    if !request.locators.contains(first_block.parent_hash()) {
-                        log::error!("Received invalid chain of missing blocks (first block's parent not in block locators)");
-                        return false;
-                    }
+                        if first_block.block_number() > request.target_block_number {
+                            log::error!(
+                                first_block = first_block.block_number(),
+                                request.target_block_number,
+                                "Received invalid chain of missing blocks (first block > target)"
+                            );
+                            return false;
+                        }
 
-                    if first_block.block_number() > request.target_block_number {
-                        log::error!(
-                            first_block = first_block.block_number(),
-                            request.target_block_number,
-                            "Received invalid chain of missing blocks (first block > target)"
-                        );
-                        return false;
+                        // If it is a macro block, also check the signatures.
+                        let last_block = blocks.last().unwrap(); // cannot be empty
+                        if last_block.is_macro() {
+                            if let Err(e) = last_block.verify_validators(&request.epoch_validators)
+                            {
+                                log::error!(
+                                    last_block = last_block.block_number(),
+                                    error = %e,
+                                    "Received invalid chain of missing blocks (macro block does not verify)"
+                                );
+                                return false;
+                            }
+                        }
```

Comment:
Is this cheaper to check than to `.hash_cached()` if that would actually hash + the associated checks there? If not then this check should remain where it was so we check cheap things which discard the block first.

Reply:
This was to keep things a bit more tidy since keeping the order means scoping the mut references too. 

Indeed the aggregation of the signatures is still more expense than the iteration over the `blocks`. I will fix it on a later commit. Thanks!